libs/server/Objects/ItemBroker/CollectionItemBroker.cs

```diff
@@ -0,0 +1,578 @@
+
+            // Add a new observer event to the event queue
+            brokerEventsQueue.Enqueue(new NewObserverEvent(observer, keys));
+
+            // Check if main loop has started, if not, start the main loop
+            if (!isStarted)
+            {
+                isStartedLock.EnterUpgradeableReadLock();
+                try
+                {
+                    if (!isStarted)
+                    {
+                        isStartedLock.EnterWriteLock();
+                        try
+                        {
+                            _ = Task.Run(Start);
+                            isStarted = true;
+                        }
+                        finally
+                        {
+                            isStartedLock.ExitWriteLock();
+                        }
+                    }
+                }
+                finally
+                {
+                    isStartedLock.ExitUpgradeableReadLock();
+                }
+            }
+
+            var timeout = timeoutInSeconds == 0
+                ? TimeSpan.FromMilliseconds(-1)
+                : TimeSpan.FromSeconds(timeoutInSeconds);
+
+            try
+            {
+                // Wait for either the result found notification or the timeout to expire
+                await observer.ResultFoundSemaphore.WaitAsync(timeout, observer.CancellationTokenSource.Token);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+
+            sessionIdToObserver.TryRemove(observer.Session.ObjectStoreSessionID, out _);
+
+            // Check if observer is still waiting for result
+            if (observer.Status == ObserverStatus.WaitingForResult)
+            {
+                // Try to set the observer result to an empty one
+                observer.HandleSetResult(CollectionItemResult.Empty);
+            }
+
+            return observer.Result;
+        }
+
+        /// <summary>
+        /// Asynchronously wait for item from collection object at srcKey and
+        /// atomically add it to collection at dstKey
+        /// </summary>
+        /// <param name="command">RESP command</param>
+        /// <param name="srcKey">Key of object to observe</param>
+        /// <param name="dstKey">Key of object to add item into</param>
+        /// <param name="session">Calling session instance</param>
+        /// <param name="timeoutInSeconds">Timeout of operation (in seconds, 0 for waiting indefinitely)</param>
+        /// <param name="cmdArgs">Additional command arguments</param>
+        /// <returns>Result of operation</returns>
+        internal async Task<CollectionItemResult> MoveCollectionItemAsync(RespCommand command, byte[] srcKey, byte[] dstKey
+            , RespServerSession session, double timeoutInSeconds, object[] cmdArgs = null)
+        {
+            var args = cmdArgs == null ? new[] { dstKey } : new object[] { dstKey }.Union(cmdArgs).ToArray();
```

Comment:
too many allocations (new) - can this be optimized to have no allocations. we will eventually move to Span<byte> for keys, until then we cannot avoid the byte[] of key, but the other allocations can be avoided.